routing/turns_generator.cpp

```diff
@@ -1062,238 +1014,242 @@ bool PathIsFakeLoop(std::vector<geometry::PointWithAltitude> const & path)
-        turnInfo.m_ingoing.m_highwayClass == turnInfo.m_outgoing.m_highwayClass &&
-        GetLinkCount(turnCandidates) + 1 == turnCandidates.size())
-    {
-      turn.m_turn = CarDirection::None;
+    /// @todo Min abs angle of ingoing ones should be considered. If it's much bigger than route angle - ignore ingoing ones.
+    /// Now this data is not available from IRoutingResult::GetPossibleTurns().
+    if (ingoingCount <= 1 + size_t(!turnInfo.m_outgoing->m_isOneWay))
       return;
-    }
   }
 
-  if (turnCandidates.size() >= 2)
-  {
-    // If the route goes along the rightmost or the leftmost way among available ones:
-    // 1. RightmostDirection or LeftmostDirection is selected
-    // 2. If the turn direction is |CarDirection::GoStraight| 
-    // and there's another GoStraight or SlightTurn turn
-    // GoStraight is corrected to TurnSlightRight/TurnSlightLeft
-    // to avoid ambiguity: 2 or more almost straight turns and GoStraight direction.
-
-    // turnCandidates are sorted by angle from leftmost to rightmost.
-    if (turnCandidates.front().m_segment == firstOutgoingSeg)
-    {
-      // The route goes along the leftmost candidate. 
-      turn.m_turn = LeftmostDirection(turnAngle);
-      // Compare with the closest left candidate.
-      GoStraightCorrection(turnCandidates[1], CarDirection::TurnSlightLeft, turn);
-    }
-    else if (turnCandidates.back().m_segment == firstOutgoingSeg)
-    {
-      // The route goes along the rightmost candidate. 
-      turn.m_turn = RightmostDirection(turnAngle);
-      // Compare with the closest right candidate.
-      GoStraightCorrection(turnCandidates[turnCandidates.size() - 2], CarDirection::TurnSlightRight, turn);
-    }
-    // Note. It's possible that |firstOutgoingSeg| is not contained in |turnCandidates|.
-    // It may happened if |firstOutgoingSeg| and candidates in |turnCandidates| are
-    // from different mwms.
-  }
+  if (CanDiscardTurnByHighwayClassOrAngles(intermediateDirection, turnAngle, turnCandidates, turnInfo, numMwmIds))
+    return;
+
+  turn.m_turn = intermediateDirection;
+
+  if (turnCandidates.size() >= 2)  
+    CorrectRightmostAndLeftmost(turnCandidates, firstOutgoingSeg, turnAngle, turn);
 }
 
 void GetTurnDirectionPedestrian(IRoutingResult const & result, size_t outgoingSegmentIndex,
                                 NumMwmIds const & numMwmIds,
                                 RoutingSettings const & vehicleSettings, TurnItem & turn)
 {
-  auto const & segments = result.GetSegments();
-  CHECK_LESS(outgoingSegmentIndex, segments.size(), ());
-  CHECK_GREATER(outgoingSegmentIndex, 0, ());
-
-  if (PathIsFakeLoop(segments[outgoingSegmentIndex].m_path))
-    return;
-
-  bool const isStartFakeLoop = PathIsFakeLoop(segments[outgoingSegmentIndex - 1].m_path);
-
-  if (isStartFakeLoop && outgoingSegmentIndex < 2)
+  TurnInfo turnInfo;
+  bool validTurnInfo = GetTurnInfo(result, outgoingSegmentIndex, vehicleSettings, turnInfo);
+  if (!validTurnInfo)
```

Comment:
if (!GetTurnInfo(...))

Reply:
Can I use https://github.com/organicmaps/organicmaps/pull/2508#issuecomment-1117040390?

Reply:
Changed to 
if (!GetTurnInfo(...))